From: Keith Busch <keith.busch@intel.com>
Date: Tue, 23 Jun 2015 14:28:34 -0600
Subject: [PATCH] NVMe: Return busy status on suspended queue
Patch-mainline: Never, in nvme-legacy repository
Git-commit: 87f745dd39adb10240b7ff14baa0dfe04b75d3b6
Git-repo: http://git.infradead.org/users/kbusch/nvme-legacy.git
References: bsc#948374

The sync path previously scheduled itself out unconditionally, but
that shouldn't happen if the command wasn't posted to the submission
queue. This patch returns immediately when the suspended queue returns
busy status.

Signed-off-by: Keith Busch <keith.busch@intel.com>
Acked-by: Johannes Thumshirn <jthumshirn@suse.de>
---
 drivers/block/nvme-core.c |    9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

--- a/drivers/block/nvme-core.c
+++ b/drivers/block/nvme-core.c
@@ -1008,7 +1008,7 @@ static int nvme_submit_sync_cmd(struct n
 						struct nvme_command *cmd,
 						u32 *result, unsigned timeout)
 {
-	int cmdid;
+	int cmdid, ret;
 	struct sync_cmd_info cmdinfo;
 	struct nvme_queue *nvmeq;
 
@@ -1027,7 +1027,12 @@ static int nvme_submit_sync_cmd(struct n
 	cmd->common.command_id = cmdid;
 
 	set_current_state(TASK_UNINTERRUPTIBLE);
-	nvme_submit_cmd(nvmeq, cmd);
+	ret = nvme_submit_cmd(nvmeq, cmd);
+	unlock_nvmeq(nvmeq);
+	if (ret) {
+		free_cmdid(nvmeq, cmdid, NULL);
+		return ret;
+	}
 	schedule();
 
 	if (result)
